Lay out loop bodies contiguously

JosephTremoulet opened the pull request:
Rearrange basic blocks during loop identification so that loop bodies
are kept contiguous when possible.  Blocks in the lexical range of the
loop which do not participate in the flow cycle (which typically
correspond to code associated with early exits using `break` or
`return`) are moved out below the loop when possible without breaking EH
region nesting.  The target insertion point, when possible, is chosen to
be the first spot below the loop that will not break up fall-through.

Layout can significantly affect the performance of loops, particularly
small search loops, by avoiding the taken branch on the hot path,
improving the locality of the code fetched while iterating the loop, and
potentially aiding loop stream detection.

Resolves #9692.

JosephTremoulet commented:
@briansull, @AndyAyersMS PTAL
/cc @dotnet/jit-contrib, @bbowyersmyth, @redknightlois

This change acheives the 30% in the microbenchmark from #9692 (added here to perf suite), as well as the 6% in [fasta](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fasta) and 5% in [TreeSort](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/BenchI/TreeSort) identified in #11192, plus 6% in [fankuch-redux](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/BenchmarksGame/fannkuch-redux), 5% in [Ackermann](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/BenchI/Ackermann), and 1% in [Puzzle](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/Performance/CodeQuality/BenchI/Puzzle).

@redknightlois, it would be great to know if this fixes the issues in your code, or if that would require improved loop detection or discontiguous EH regions or something.

Throughput impact negligible (0.07% instrs retired increased in release System.Private.CoreLib crossgen).

Results from jit-diff:

```
Summary:
(Note: Lower is better)

Total bytes of diff: -6622 (0.00 % of base)
    diff is an improvement.

Total byte diff includes -500 bytes from reconciling methods
        Base had    1 unique methods,      500 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file regressions by size (bytes):
         184 : System.Net.Security.dasm (0.13 % of base)
         148 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.01 % of base)
         118 : Microsoft.CSharp.dasm (0.04 % of base)
          47 : System.Linq.dasm (0.03 % of base)
          40 : JIT\Performance\CodeQuality\BenchF\MatInv4\MatInv4\MatInv4.dasm (0.99 % of base)

Top file improvements by size (bytes):
       -1710 : System.Private.CoreLib.dasm (-0.05 % of base)
        -719 : Microsoft.CodeAnalysis.CSharp.dasm (-0.03 % of base)
        -204 : JIT\jit64\opt\cse\staticFieldExpr1_ro_loop\staticFieldExpr1_ro_loop.dasm (-11.45 % of base)
        -187 : System.Private.Uri.dasm (-0.24 % of base)
        -172 : Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest\MarshalArrayByValTest.dasm (-1.05 % of base)

344 total files with size differences (299 improved, 45 regressed), 6187 unchanged.

Top method regessions by size (bytes):
         296 : System.Private.CoreLib.dasm - Type:GetEnumData(byref,byref):this
         196 : System.Net.Security.dasm - SSPIWrapper:EncryptDecryptHelper(int,ref,ref,ref,int):int
         110 : System.Private.CoreLib.dasm - GenericEqualityComparer`1:IndexOf(ref,ref,int,int):int:this
          89 : System.Private.CoreLib.dasm - EventProvider:WriteEvent(byref,long,long,long,ref):bool:this
          75 : Microsoft.CodeAnalysis.VisualBasic.dasm - SourceModuleSymbol:BindImports(struct):ref:this

Top method improvements by size (bytes):
        -500 : Microsoft.CodeAnalysis.CSharp.dasm - LanguageParser:SkipBadTokensWithExpectedKind(ref,ref,char,byref):int:this
        -204 : JIT\jit64\opt\cse\staticFieldExpr1_ro_loop\staticFieldExpr1_ro_loop.dasm - Test_Main:Main():int
        -183 : System.Private.CoreLib.dasm - Dictionary`2:Remove(ref):bool:this (11 methods)
        -172 : Interop\ArrayMarshalling\ByValArray\MarshalArrayByValTest\MarshalArrayByValTest.dasm - Test:Equals(ref,ref):bool (11 methods)
        -129 : System.Private.CoreLib.dasm - Enum:TryParseEnum(ref,ref,bool,byref):bool

1632 total methods with size differences (1096 improved, 536 regressed), 313322 unchanged.
```

redknightlois commented:
@JosephTremoulet I have the right benchmark for those :) ... We will be moving the codebase to 2.0 by next week (hopefully) so I could be able to try this. Loop code flow has given us massive improvements, so even if you dont catch all versions it should be a net win performance wise for tight code.

stephentoub commented:
@JosephTremoulet, if this fixes the problem, it'd be great to also undo the workarounds employed in a variety of places, e.g.
https://github.com/dotnet/coreclr/blob/57214f1948e56835cde1b6162acc58751f483e14/src/mscorlib/src/System/String.Comparison.cs#L47
https://github.com/dotnet/corefx/blob/b769a73ecf278f92993a8b789aaaaedbc55a7dc3/src/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs#L334
https://github.com/dotnet/corefx/blob/cab2165f92380e8424589909158a26ce79e18683/src/System.Memory/src/System/SpanHelpers.byte.cs#L158
(I think there are also some cases where a comment wasn't left, so it might be worth searching for gotos.)

BruceForstall commented:
@JosephTremoulet this change calls for full desktop testing plus tagging dotnet-bot to run a bunch of stress modes.

JosephTremoulet commented:
@stephentoub sounds good, I'll verify and revert as the compiler change propagates

JosephTremoulet commented:
@dotnet-bot help

JosephTremoulet commented:
@dotnet-bot  test Windows_NT x86 Release gcstress15_pri1r2r  test Windows_NT x86 corefx_jitstressregs3  test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc  test Ubuntu jitstress2_jitstressregs1 

@BruceForstall feel free to add if I missed any important ones

AndyAyersMS commented on src/jit/optimizer.cpp:
```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
+                        // This code is lexically between TOP and BOTTOM, but it does not
+                        // participate in the flow cycle.  Check for a run of consecutive
+                        // such blocks.
+                        BasicBlock* lastExitBlock = block;
```
`lastExitBlock` sounds a lot like a name for a block that is in a loop. Maybe `lastNonLoopBlock`?

JosephTremoulet replied:
Done

AndyAyersMS commented on src/jit/optimizer.cpp:
```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
```
I wonder if this would be a clearer if it was conceptually split up as follows (with appropriate bits of state passed in and out):
```
for block in range [top..bottom]
   previous = block->bbPrev;
   if block is in loop: CountLoopExit(...); next = block->bbNext
   if block is not in loop: next = MoveBlockRangeOutOfLoopOrTolerateInLoop(...)  // ret null if fails
```

JosephTremoulet replied:
I agree the function is more monolithic than ideal, so I'll see if I can find an appropriate way to refactor, but at first blush I don't think it would end up reading better, given the amount of state that would need to get passed in and particularly that would need to get passed out, not to mention I'd need to pull the `LoopBlockSet` definition out of line.

JosephTremoulet replied:
> > I wonder if this would be a clearer if it was conceptually split up...

> I agree the function is more monolithic than ideal...  the amount of state that would need to get passed in and particularly that would need to get passed out,..

Of course, I could wrap that state up in a class and pass it around implicitly.  @AndyAyersMS, let me know if you think https://github.com/JosephTremoulet/coreclr/commit/0861de1cb4efc026cf533759fa8c735938a9adbe is an improvement and I can clean it up and push it to the PR branch.

AndyAyersMS replied:
Thanks for considering this -- I think it is an improvement and really helps clarify what is going on.

AndyAyersMS commented on src/jit/optimizer.cpp:
```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
+
+                        if (moveAfter->bbFallsThrough())
+                        {
+                            // We've just inserted blocks between moveAfter and moveBefore which it was supposed
```
Would be clearer to me if `moveBefore` was actually a thing here. It took me a while to figure out that `lastExitBlock->bbNext` was the block that `moveAfter` falls through to originally.

JosephTremoulet replied:
Good point, defined `moveBefore` in the update.

JosephTremoulet commented on src/jit/optimizer.cpp:
```diff
@@ -1574,177 +1586,149 @@ void Compiler::optFindNaturalLoops()
-                        }
-                    }
+                    // The "back-edge" we identified isn't actually part of the flow cycle containing ENTRY
+                    goto NO_LOOP;
```
You're following correctly.  Our loop detection algorithm (both before and after this change) already walks all inclusive blocks of each loop and is hence quadratic, regardless of this bail-out.  We cap the number of loops we track at 16, which I thought was what prevented this from running away; I've just now discovered that the code didn't actually bail out but instead kept looking for loops that it was doomed to discard, so I've added a bail-out in the update I just pushed.

Regarding "Do we need a similar check for TOP? Seems like we want to ensure that TOP and BOTTOM define the lexical extent of the loop so that the range-inclusive checks later work out" -- yes, we did, thanks.  I was thinking that finding BOTTOM in the set was sufficient because it has TOP as a successor, but I had that backwards because the walk is visiting predecessors -- so finding TOP is sufficient because it has BOTTOM as a predecessor, and I've pushed an update accordingly.

JosephTremoulet commented on src/jit/optimizer.cpp:
```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
+                                    if ((destNum >= top->bbNum) && (destNum <= bottom->bbNum) &&
+                                        !loopBlocks.isMember(destNum))
+                                    {
+                                        // Reversing this branch out of block `next` could confuse this algorithm, so
```
Here we're not happy that `moveAfter` splits up fall-through, so we're checking to see if `next` is a viable candidate to become our `moveAfter` -- in which case we would be inserting blocks after it.  So I believe the code is correct but hard to follow.  I've added a comment up where `next` is defined to try to clarify a bit, let me know if you think something else would help.

AndyAyersMS replied:
Ah, you have an invariant that `moveAfter` is always a viable (if not optimal) place to move the blocks.

Maybe the other `break` cases confused me -- presumably you have a range of candidate blocks to insert after that starts with BOTTOM and extends all the way to the end of the EH region that contains the blocks you'd like to move. 

So for these "reversing would confuse" cases why wouldn't you keep going instead of breaking the search there and accepting the current `moveAfter`? And if just after BOTTOM is a viable point then presumably you could just search from there to the end for a non-fallthrough and if you find one, use it, otherwise just insert after BOTTOM.

Also note your search loop doesn't have to worry about walking off the end of the method since the last block is guaranteed to be non-fallthrough. So if you are not in an EH region you might just move the blocks to the end of the method and skip the searching all together.

JosephTremoulet replied:
Those branches would still be reversed if we moved the blocks anywhere after `next`.

The clause where we don't search into a different EH region is one where we could later find another viable candidate (if the blocks after the loop fall through into a new `try` and then back out of it), but I'm not sure it's worth the complexity to keep searching for those cases.

Regarding "if you find one, use it, otherwise just insert after BOTTOM", were you just sketching out how an algorithm that kept walking would work, or do you have a reason to think that BOTTOM is a better insertion point than the last legal block?

AndyAyersMS replied:
> Those branches would still be reversed if we moved the blocks anywhere after next.

That's the bit I'm missing. If the code can find a fallthrough, then the "fixup" for the in loop blocks is something it can already handle, so I don't see yet how skipping past a problematic next to some non-fallthrough spot below it wouldn't work. Maybe you just need to be a bit more explicit on what exactly about reversal leads to confusion.

> Regarding "if you find one, use it, otherwise just insert after BOTTOM", were you just sketching out how an algorithm that kept walking would work, or do you have a reason to think that BOTTOM is a better insertion point than the last legal block?

Just thinking how I would have started writing this bit. I probably would have tried to leverage `fgFindInsertPoint` to do the actual searching, and looking at it now, it seems like there might be logic there you need here, eg a BBJ_CALLFINALLY does not fall through but you can't safely move new blocks after it.

As far as a "best" insertion point -- not creating a new block is preferable from a TP standpoint. Other than that I would say try and avoid creating new lexically backwards branches. Insert after BOTTOM would guarantee this, though you could get the same guarantee if you moved the run of blocks before any post-loop block they reach, and if there is just one such block and maybe try and move the run just before it.

I suppose it would also be nice to avoid moving the blocks into another loop where you'll probably just end up moving them again later on, but that might be tricky to pull off or maybe doesn't happen with any frequency. You could perhaps use `BBF_BACKWARD_JUMP` as a screen.

None of this is all that important, if what you have now is correctly moving blocks. If so I would not worry as much about trying to make it optimal, unless you already see issues that come from particular placement choices.

JosephTremoulet replied:
> Maybe you just need to be a bit more explicit on what exactly about reversal leads to confusion.

The algorithm expects that if it can find an edge from a block with higher number to a block with lower number, that it has found a lexically backward edge and therefore that walking `bbNext` pointers from the successor will find the predecessor.  It also assumes, when walking the loop blocks, that any block it visits with number between TOP and BOTTOM is lexically between the two. 
 Performing any motion that reverses the lexical direction of a flow edge risks breaking those invariants.
(and the motion we do perform can mean that blocks lexically between TOP and BOTTOM in a subsequent loop have numbers less than TOP's number, but I believe that can only cause conservativism [we'll treat flow into or out of that region as exits and side-entries when we maybe didn't need to], not correctness issues)

> avoid creating new lexically backwards branches

I think that's a correctness constraint, and that's what this code is trying to do.

> Insert after BOTTOM would guarantee this, though ...

I started with always inserting just after bottom, but the diffs showed lots of cases where that led to ugly layout; in particular, given a small method with a search loop that has an early return and the normal loop exit falls through to the other return, it unnecessarily puts an unconditional jump on the normal exit path.

> you could get the same guarantee if you moved the run of blocks before any post-loop block they reach, and if there is just one such block and maybe try and move the run just before it.

That's really what this code is trying to do.  And since I can't guarantee correctness if I move past any such block, the search stops at the first such block and we move the run just before it.


> I probably would have tried to leverage `fgFindInsertPoint` to do the actual searching, and looking at it now, it seems like there might be logic there you need here, eg a `BBJ_CALLFINALLY` does not fall through but you can't safely move new blocks after it.

Thanks, I didn't know about `fgFindInsertPoint` or that constraint; I'll have to take a look and update...

AndyAyersMS replied:
The other thought I've had: defer the actual motion to a post-pass, and just identify the set of blocks that do not belong and can be moved out. I suppose you might also need to simulate moving them...

JosephTremoulet replied:
I looked through `fgFindInsertPoint` and think the Call/Always pair check is the only thing to bring over, so have pushed an update that includes it.

> defer the actual motion to a post-pass, and just identify the set of blocks that do not belong and can be moved out. I suppose you might also need to simulate moving them...

I considered that, but wanted to first see if it was tenable to do the in-place transform to avoid the extra heap traffic, and I'm happy with the result.  And I agree, the "simulate moving them" bookkeeping might end up just as hairy.

JosephTremoulet commented on src/jit/optimizer.cpp:
```diff
@@ -1483,12 +1483,17 @@ void Compiler::optFindNaturalLoops()
     BasicBlock*   top;
     BasicBlock*   bottom;
     BasicBlock*   entry;
     BasicBlock*   exit;
```
Done.  I separated `lastExit` and `onlyExit` while at it, which makes it read more clearly I think.

JosephTremoulet commented on src/jit/optimizer.cpp:
```diff
@@ -1574,177 +1586,149 @@ void Compiler::optFindNaturalLoops()
-                            exitPoint = loopBlock->bbJumpDest;
+                        if (blockNum > oldBlockMaxNum)
+                        {
+                            BlockSetOps::AddElemD(comp, newBlocksInLoop, blockNum - oldBlockMaxNum);
```
Added.

JosephTremoulet commented on src/jit/optimizer.cpp:
```diff
@@ -1574,177 +1586,149 @@ void Compiler::optFindNaturalLoops()
                 {
-                    if (predEntry->flBlock->bbNum >= top->bbNum && predEntry->flBlock->bbNum <= bottom->bbNum)
+                    BasicBlock* block = worklist.back();
+                    worklist.pop_back();
```
It can; good catch, thanks.  Updated, but rather than using `positionNum` (since `fgDominate` takes the blocks), I just defer the check until processing the predecessor.

JosephTremoulet commented on tests/src/JIT/Regression/JitBlue/GitHub_9692/GitHub_9692.cs:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        // Can't move out of loop without crossing try region boundary; should leave in loop
+        // Lopo should still be recognized, and invariant multiplication hoisted out of it.
```
Fixed.

JosephTremoulet commented:
I pushed an update to fix an issue found in desktop testing where creating a goto-next caused an assertion failure.  If desktop and normal CI testing are green with the fix, I'll squash the updates and re-trigger stress runs.

AndyAyersMS commented on src/jit/optimizer.cpp:
```diff
@@ -1574,179 +1588,165 @@ void Compiler::optFindNaturalLoops()
+                                continue;
+                            }
+
+                            // There are multiple entries to this loop, don't consider it.
```
This "side entry" could also be from a disjoint part of the loop. It might be interesting someday to work through relaxing this condition. Something like: let the pred closure happen, since it seems like the new dominates check should prevent wandering outside the loop, and see how many disjoint bodies there are out there.

If these are common, we may now have the machinery to compact them too.

JosephTremoulet commented:
Updated to include refactoring.

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 Release gcstress15_pri1r2r test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc test Ubuntu jitstress2_jitstressregs1

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 Release gcstress15_pri1r2r test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc test Ubuntu jitstress2_jitstressregs1

JosephTremoulet commented:
@dotnet-bot re-test OSX10.12 x64 Checked Build and Test ([infrastructure failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/checked_osx10.12_tst_prtest/4837/consoleFull#16952254646ee26338-6221-4f7c-8c8f-1d1d5cb4704a))

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 Release gcstress15_pri1r2r test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc test Ubuntu jitstress2_jitstressregs1

briansull commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,978 @@ void Compiler::optCheckPreds()
+//             but not the outer loop. ???)
+//   TOP     - the target of the backward edge from BOTTOM. In most cases FIRST and TOP are the same.
+//   BOTTOM  - the lexically last block in the loop (i.e. the block from which we jump to the top)
+//   EXIT    - the predecessor of loop's unique exit edge, if it has a unique exit edge; else nullptr
```
Should this be changed to LastExit?

briansull commented on src/jit/compiler.h:
```diff
@@ -5314,6 +5314,14 @@ class Compiler
+                       BasicBlock*   top,
+                       BasicBlock*   entry,
+                       BasicBlock*   bottom,
+                       BasicBlock*   exit,
```
Should this be changed to lastexit?

briansull commented on src/jit/optimizer.cpp:
```diff
@@ -1119,10 +1119,11 @@ bool Compiler::optExtractInitTestIncr(
  */
 
-void Compiler::optRecordLoop(BasicBlock*   head,
+bool Compiler::optRecordLoop(BasicBlock*   head,
```
exit => lastexit?

JosephTremoulet replied:
It shouldn't be `lastExit` because that is just something we track while identifying the loop and then throw it away if there were multiple exits.  The code that checks for multiples has a local called `onlyExit` that it populates iff `lastExit` was unique.  So if we want to change the name of this parameter and the comments describing loop structure and the `lpExit` field in the loop table, we could change it to `onlyExit` or perhaps better `uniqueExit`, but that seems unnecessarily verbose to me...

briansull replied:
ok

JosephTremoulet commented:
I pushed a fix for an issue discovered by stress testing.

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 Release gcstress15_pri1r2r test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc test Ubuntu jitstress2_jitstressregs1

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+//   HEAD    - the basic block that flows into the loop ENTRY block (Currently MUST be lexically before entry).
+//             Not part of the looping of the loop.
+//   FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
+//             but not the outer loop. ???)
```
>???) [](start = 39, length = 4)

You copied the "???" from before, but since you've been working on this code: do you know the answer?

JosephTremoulet replied:
I don't understand the question.  Is the question whether `fist` may be part of "a nested loop", and also whether it may be part of "the outer loop"?  If so, "nested" and "outer" with respect to what?  Why is it "**a** nested loop" vs. "**the** outer loop"?  If we're talking about the full loop nest forest, then there aren't any restrictions on how deeply nested a loop with a "first" is.  If we're talking about whether a given loop's "first" may be included in any "outer" loops enclosing the given one, then yes of course it is a member of all of them.  If we're talking about whether a given loop's "first" may be included in any "nested" loops enclosed by the given one, then:
 - as far as `LoopSearch` is concerned, there's nothing conceptually barring that
 - there's code called after `loopSearch`, but still within `optFindNaturalLoops`, that canonicalizes loops; its comments say the point is to have no two loops share `top` but then the implementation talks about sharing `fist`
 - in the current implementation, `first` and `top` are always the same, and the search algorithm in `LoopSearch` will never find two loops that share one

So I don't know what was supposed to be said here.  Should I just remove the whole parenthetical?

BruceForstall replied:
> Should I just remove the whole parenthetical?

I'd be ok with that.

Perhaps a discussion somehere (here?) about the relationship between all the concepts here (FIRST/TOP/etc.) and nested loops. e.g., can a FIRST be the FIRST of multiple loops?

JosephTremoulet replied:
Clarified comments in https://github.com/dotnet/coreclr/pull/13314/commits/73f116794e24aa5517dca28914ce2a7889634539

JosephTremoulet replied:
Thanks for prompting me to take a closer look at this.  I had misread and thought the old code wouldn't identify loops that shared a `TOP` or `ENTRY` or `FIRST` -- on closer read, it allowed it as long as `entry`'s predecessor list had an inner-loop predecessor prior to having an outer-loop predecessor (oddly).  So I've fixed my code to recognize nested loops that share `TOP`/`FIRST`/`ENTRY` in https://github.com/dotnet/coreclr/pull/13314/commits/13bcb530f4f66ca9bdd006fab7a2d7c7f3040c97 (before refactoring) and https://github.com/dotnet/coreclr/pull/13314/commits/1091e8947ac147d2a900aa0055909402566c5704 + https://github.com/dotnet/coreclr/pull/13314/commits/a9c20d2ec763186650e1c0ff49c572760e5d7525 (after refactoring), without the bizarro restriction on predecessor list.

AndyAyersMS replied:
Yeah, this is the whole "keep the pred list sorted" thing that came up over in #13322. If loop recognition no longer needs this then maybe we can drop the whole thing.

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+//        v
+//      head
+//        |
+//        |    top/beg <--+
```
Should "beg" in "top/beg" be "first"? Should there be an example where TOP and FIRST are not the same?

JosephTremoulet replied:
Probably should be "first".  We do not currently have any examples where TOP and FIRST are not the same.

JosephTremoulet replied:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/73f116794e24aa5517dca28914ce2a7889634539

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+//             Not part of the looping of the loop.
+//   FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
+//             but not the outer loop. ???)
+//   TOP     - the target of the backward edge from BOTTOM. In most cases FIRST and TOP are the same.
```
>In most cases FIRST and TOP are the same. [](start = 60, length = 41)

"In most cases FIRST and TOP are the same." It would be useful if the comment gave an example of when they are NOT the same.

JosephTremoulet replied:
Yes, I think that would be a useful comment to add if/when we implement support for that.

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 Release gcstress15_pri1r2r test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Windows_NT gcstress0xc test Ubuntu jitstress2_jitstressregs1

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+            assert(seedBlockNum <= oldBlockMaxNum);
+            oldBlocksInLoop = BlockSetOps::MakeSingleton(comp, seedBlockNum);
+            newBlocksInLoop = BlockSetOps::MakeEmpty(comp);
+        }
```
These will allocate new bitsets. Why not use ClearD() and reuse the existing allocation? (Make the constructor do the initial allocations?)

JosephTremoulet replied:
Thanks for catching that, I was under the mistaken impression that the LongRep was sparse.  Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/f86557bbcb5fc77ea6b190a7769d39ee09a98305

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+        bool isMember(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
```
All of these functions need
```
noway_assert(blockNum - oldBlockMaxNum < oldBlockMaxNum);
```
or perhaps:
```
noway_assert(blockNum - oldBlockMaxNum < BasicBlockBitSetTraits::GetSize(comp));
```
in the "new blocks" cases, to ensure we don't allocate more than 1 new block per original block.

JosephTremoulet replied:
is there a reason you want it to be a `noway_assert` instead of a plain `assert`?

BruceForstall replied:
plain would be ok, i guess

---
In reply to: [133579212](https://github.com/dotnet/coreclr/pull/13314#discussion_r133579212) [](ancestors = 133579212)

JosephTremoulet replied:
Added `noway_assert` where the new blocks are created in https://github.com/dotnet/coreclr/pull/13314/commits/0b85921eda261a713479b4722d8542363982369b

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    BlockSet      bottomBlocks;   // BOTTOM blocks of already-recorded loops
+#ifdef DEBUG
+    bool forgotExit = false; // Flags a rare case where lastExit gets nulled out, for assertions
+#endif
```
Does C++ translate this assignment into a ctor initializer?

JosephTremoulet replied:
Yes.

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+
+public:
+    bool changedFlowGraph = false; // Signals to caller that fgUpdateChangedFlowGraph needs to be called
+
```
shouldn't this be private, with public accessor?

JosephTremoulet replied:
Why?

BruceForstall replied:
Do the class users both set and read it? If the class sets it and users read it, it would be safer for the users to not have write access.

JosephTremoulet replied:
Ok; privatized field in https://github.com/dotnet/coreclr/pull/13314/commits/72f1eea5c673ea616144271a21fed451932f39d3

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
+    //
+    bool findLoop(BasicBlock* head)
```
CC (nit?): member func names should be PascalCase [https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#8.9](https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#8.9). (We're probably not too consistent here currently)

JosephTremoulet replied:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/71989abf833df501ccbad9fb5ff42b6693749581

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    //    false - Did not find a valid loop.
+    //
+    // Notes:
+    //    May modify flow graph go make loop compact before returning.
```
>go [](start = 32, length = 2)

go => to?

JosephTremoulet replied:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/29ce6e11b8cf5112b4ffca424e96441dfe7b4f10

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    //
+    // Notes:
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
```
>Will set instance fields to track loop's extent and exits. [](start = 10, length = 58)

... if a valid loop is found

JosephTremoulet replied:
... or not

JosephTremoulet replied:
Expanded in https://github.com/dotnet/coreclr/pull/13314/commits/29ce6e11b8cf5112b4ffca424e96441dfe7b4f10

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
+    //
+    bool findLoop(BasicBlock* head, BasicBlock* top, BasicBlock* bottom)
```
Can't this have a different name from the "top-level" findLoop?

JosephTremoulet replied:
Of course it can, but why and what would you prefer?

BruceForstall replied:
why: they do different work; easier to "grep". prefer: FindLoopFromCandidateBottom()? FindLoopWorker?

---
In reply to: [133580782](https://github.com/dotnet/coreclr/pull/13314#discussion_r133580782) [](ancestors = 133580782)

JosephTremoulet replied:
No longer have two methods as of https://github.com/dotnet/coreclr/pull/13314/commits/a9c20d2ec763186650e1c0ff49c572760e5d7525

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+    //    oldMoveAfter - Prior insertion point; find the next after this.
+    //
+    // Return Value:
+    //    The next block after `oldmoveAfter` that is a legal insertion point
```
nit: captitalization: `oldmoveAfter` => `oldMoveAfter`

JosephTremoulet replied:
Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/29ce6e11b8cf5112b4ffca424e96441dfe7b4f10

BruceForstall commented on src/jit/optimizer.cpp:
```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+
+        // Unable to record this loop because the loop descriptor table overflowed.
+        return false;
+    }
```
If the loop table is full, wouldn't we know that before we did all the work to detect the loop? I.e., why do all that work if we're just going to fail to add it to the loop table?

JosephTremoulet replied:
It seemed odd to me to, but we have code to [recognize](https://github.com/dotnet/coreclr/blob/d445122aad3b7c542f77b0b031cb9f205054a5c9/src/jit/optimizer.cpp#L1139) when loops are being dropped on the floor and [count](https://github.com/dotnet/coreclr/blob/d445122aad3b7c542f77b0b031cb9f205054a5c9/src/jit/optimizer.cpp#L1832-L1835) the frequency, so I was just trying to preserve that here.

JosephTremoulet replied:
I went back and took a closer look.  We compile with `COUNT_LOOPS` `#define`d to `0` by default.  Also, the prior behavior was to never bail out of searching for loops, that was something I was adding in this change (per some [earlier feedback](https://github.com/dotnet/coreclr/pull/13314#discussion_r132538713)).  So I've rearranged things to count all loops when `COUNT_LOOPS` is `1` (as before), but to bail out before finding a doomed loop otherwise.  Fixed in https://github.com/dotnet/coreclr/pull/13314/commits/6d7fcf6d55fcad8b765c14ba215a9f04cd52a4e3 (before refactoring) and https://github.com/dotnet/coreclr/pull/13314/commits/2feecd675c62098c876d88cce7a84119da61b086 (after refactoring)

BruceForstall commented:
Is it possible to run tests + diffs with your refactoring, but without your functional changes, to ensure no diffs, and then have a separate change to enable your block reorganization changes? This would allow the changes to go through all our internal testing separately.

JosephTremoulet commented:
> Is it possible to run tests + diffs with your refactoring, but without your functional changes, to ensure no diffs, and then have a separate change to enable your block reorganization changes? This would allow the changes to go through all our internal testing separately.

I got desktop testing green before adding the refactoring, and ran jit-diff with the change before and after the refactoring (no diffs).  Changes made since have been minimal.

JosephTremoulet commented:
The ubuntu arm64 [failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size_prtest/689/) matches the [baseline](https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size/507/)

The GCStress-0xC [failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc_prtest/28/) likewise matches the [baseline](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc/48)

JosephTremoulet commented:
x86 GCStress 15 R2R [failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_release_windows_nt_gcstress15_pri1r2r_prtest/6/) matches [baseline](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_release_windows_nt_gcstress15_pri1r2r/32/) as well

BruceForstall commented:
Have you tried a throughput job? Or looked at memory usage diff?

JosephTremoulet commented:
TP: I measured 0.07% instrs retired increased in release System.Private.CoreLib crossgen
Memory: pending switching to `ClearD`, the heap allocations are three BlockSet BVs and one DFS worklist, but sure I can verify

JosephTremoulet commented:
@BruceForstall, I think I've covered all your feedback (except running mem diff tests); PTAL.  I back-ported any behavioral changes so that we can still have separate commits for the functionality change and the refactoring; if the changes look good I'll squash the various fixes so just those two commits remain, but for now I kept them separate to allow differential review.

I made some functional changes regarding nested loops while addressing the feedback, so I'll kick off the various test passes again.

JosephTremoulet commented:
@dotnet-bot test Windows_NT x86 corefx_jitstressregs3 test Windows_NT x86 Checked jitstress2 test Windows_NT Checked r2r_jitstressregs8 test Windows_NT jitstress1 test Ubuntu jitstress2_jitstressregs1

JosephTremoulet commented:
Stress [failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_corefx_jitstressregs3_prtest/10/) matches [baseline](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_corefx_jitstressregs3/226/)

JosephTremoulet commented:
Memstats reports 0.1% more allocated bytes:

<details>

```diff
diff --git "a/memstats-base.txt" "b/memstats-diff.txt"
index baa2dee93..9650c0d99 100644
--- "a/memstats-base.txt"
+++ "b/memstats-diff.txt"
@@ -1,133 +1,133 @@
 
 D:\Source\coreclr>D:\Source\coreclr\bin\Product\Windows_NT.x64.Release\crossgen.exe /Platform_Assemblies_Paths D:\Source\coreclr\bin\Product\Windows_NT.x64.Release\IL /out D:\Source\coreclr\bin\Product\Windows_NT.x64.Release\System.Private.CoreLib.dll D:\Source\coreclr\bin\Product\Windows_NT.x64.Release\IL\System.Private.CoreLib.dll 
 
 All allocations:
 For     26291 methods:
-  count:           20198079 (avg     768 per method)
+  count:           20269450 (avg     770 per method)
-  alloc size :   1813175957 (avg   68965 per method)
+  alloc size :   1815665942 (avg   69060 per method)
   max alloc  :        86912
 
-  allocateMemory   :   3336175616 (avg  126894 per method)
+  allocateMemory   :   3338403840 (avg  126978 per method)
-  nraUsed    :   2546966432 (avg   96875 per method)
+  nraUsed    :   2549453232 (avg   96970 per method)
 
 Alloc'd bytes by kind:
                   kind |       size |     pct
   ---------------------+------------+--------
-         AssertionProp |  170463080 |   9.40%
+         AssertionProp |  170463080 |   9.39%
-               ASTNode |  332156679 |  18.32%
+               ASTNode |  332244447 |  18.30%
-              InstDesc |   49788288 |   2.75%
+              InstDesc |   49833836 |   2.74%
               ImpStack |     498576 |   0.03%
-            BasicBlock |   59676128 |   3.29%
+            BasicBlock |   59727608 |   3.29%
-             fgArgInfo |   10293192 |   0.57%
+             fgArgInfo |   10293752 |   0.57%
-       fgArgInfoPtrArr |    1470456 |   0.08%
+       fgArgInfoPtrArr |    1470536 |   0.08%
-              FlowList |    7445776 |   0.41%
+              FlowList |    8109200 |   0.45%
-     TreeStatementList |    1820768 |   0.10%
+     TreeStatementList |    1823072 |   0.10%
-               SiScope |    8998360 |   0.50%
+               SiScope |    8832344 |   0.49%
         FlatFPStateX87 |          0 |   0.00%
-       DominatorMemory |    5762000 |   0.32%
+       DominatorMemory |    6364728 |   0.35%
-                  LSRA |   88207680 |   4.86%
+                  LSRA |   88227684 |   4.86%
-         LSRA_Interval |   55517616 |   3.06%
+         LSRA_Interval |   55520168 |   3.06%
-      LSRA_RefPosition |  171685080 |   9.47%
+      LSRA_RefPosition |  171686480 |   9.46%
           Reachability |     504992 |   0.03%
-                   SSA |   43350424 |   2.39%
+                   SSA |   43346104 |   2.39%
-           ValueNumber |  338346518 |  18.66%
+           ValueNumber |  338356155 |  18.64%
-              LvaTable |   82502848 |   4.55%
+              LvaTable |   82502872 |   4.54%
-            UnwindInfo |     191968 |   0.01%
+            UnwindInfo |     191744 |   0.01%
-                hashBv |    5633656 |   0.31%
+                hashBv |    5633568 |   0.31%
-                bitset |   28625144 |   1.58%
+                bitset |   28714696 |   1.58%
           FixedBitVect |       1056 |   0.00%
-          AsIAllocator |  150184316 |   8.28%
+          AsIAllocator |  150910712 |   8.31%
         IndirAssignMap |      85328 |   0.00%
          FieldSeqStore |    4782488 |   0.26%
     ZeroOffsetFieldMap |    1283880 |   0.07%
-          ArrayInfoMap |    2489776 |   0.14%
+          ArrayInfoMap |    2490016 |   0.14%
-          MemoryPhiArg |     945952 |   0.05%
+          MemoryPhiArg |     946720 |   0.05%
-                   CSE |   38697824 |   2.13%
+                   CSE |   38698800 |   2.13%
-                    GC |   65404271 |   3.61%
+                    GC |   65393663 |   3.60%
                 CorSig |    6770920 |   0.37%
-              Inlining |   36225376 |   2.00%
+              Inlining |   36172960 |   1.99%
-            ArrayStack |   13125504 |   0.72%
+            ArrayStack |   13500672 |   0.74%
-             DebugInfo |    9907920 |   0.55%
+             DebugInfo |    9911400 |   0.55%
              DebugOnly |          0 |   0.00%
                Codegen |          0 |   0.00%
-               LoopOpt |      67200 |   0.00%
+               LoopOpt |      67440 |   0.00%
-             LoopHoist |    1586920 |   0.09%
+             LoopHoist |    1621912 |   0.09%
-               Unknown |   18677997 |   1.03%
+               Unknown |   18682333 |   1.03%
 
 
 Largest method:
-count:      78538, size:    4085578, max =      59992
+count:      80075, size:    4146298, max =      59992
-allocateMemory:    4259840, nraUsed:    4153976
+allocateMemory:    4325376, nraUsed:    4214760
 
 Alloc'd bytes by kind:
                   kind |       size |     pct
   ---------------------+------------+--------
          AssertionProp |       6460 |   0.16%
-               ASTNode |    1033792 |  25.30%
+               ASTNode |    1035328 |  24.97%
-              InstDesc |      68804 |   1.68%
+              InstDesc |      68804 |   1.66%
               ImpStack |        192 |   0.00%
-            BasicBlock |      88272 |   2.16%
+            BasicBlock |      90384 |   2.18%
-             fgArgInfo |      22064 |   0.54%
+             fgArgInfo |      22064 |   0.53%
        fgArgInfoPtrArr |       3152 |   0.08%
-              FlowList |      62640 |   1.53%
+              FlowList |      73232 |   1.77%
-     TreeStatementList |      10528 |   0.26%
+     TreeStatementList |      10528 |   0.25%
                SiScope |       8960 |   0.22%
         FlatFPStateX87 |          0 |   0.00%
-       DominatorMemory |      41120 |   1.01%
+       DominatorMemory |      48048 |   1.16%
-                  LSRA |     114708 |   2.81%
+                  LSRA |     117876 |   2.84%
-         LSRA_Interval |     141944 |   3.47%
+         LSRA_Interval |     141944 |   3.42%
-      LSRA_RefPosition |     446656 |  10.93%
+      LSRA_RefPosition |     446656 |  10.77%
           Reachability |         32 |   0.00%
-                   SSA |      55236 |   1.35%
+                   SSA |      55428 |   1.34%
-           ValueNumber |     342067 |   8.37%
+           ValueNumber |     342491 |   8.26%
-              LvaTable |      82776 |   2.03%
+              LvaTable |      82776 |   2.00%
             UnwindInfo |         32 |   0.00%
                 hashBv |       5400 |   0.13%
-                bitset |     635680 |  15.56%
+                bitset |     644168 |  15.54%
           FixedBitVect |          0 |   0.00%
-          AsIAllocator |     660700 |  16.17%
+          AsIAllocator |     685644 |  16.54%
         IndirAssignMap |          0 |   0.00%
          FieldSeqStore |        440 |   0.01%
     ZeroOffsetFieldMap |         64 |   0.00%
-          ArrayInfoMap |      15800 |   0.39%
+          ArrayInfoMap |      15800 |   0.38%
-          MemoryPhiArg |       3184 |   0.08%
+          MemoryPhiArg |       3216 |   0.08%
                    CSE |      18528 |   0.45%
-                    GC |      96796 |   2.37%
+                    GC |      96796 |   2.33%
                 CorSig |       8528 |   0.21%
-              Inlining |      25680 |   0.63%
+              Inlining |      25680 |   0.62%
-            ArrayStack |      40448 |   0.99%
+            ArrayStack |      42624 |   1.03%
              DebugInfo |       9576 |   0.23%
              DebugOnly |          0 |   0.00%
                Codegen |          0 |   0.00%
-               LoopOpt |       4320 |   0.11%
+               LoopOpt |       4320 |   0.10%
              LoopHoist |       7928 |   0.19%
-               Unknown |      23071 |   0.56%
+               Unknown |      23199 |   0.56%
 
 
 ---------------------------------------------------
 Distribution of total memory allocated per method (in KB):
      <=         20 ===>       0 count (  0% of total)
      21 ..      50 ===>       0 count (  0% of total)
      51 ..      75 ===>   11123 count ( 42% of total)
      76 ..     100 ===>       0 count ( 42% of total)
-    101 ..     150 ===>   11091 count ( 84% of total)
+    101 ..     150 ===>   11079 count ( 84% of total)
-    151 ..     250 ===>    2400 count ( 93% of total)
+    151 ..     250 ===>    2407 count ( 93% of total)
-    251 ..     500 ===>    1455 count ( 99% of total)
+    251 ..     500 ===>    1459 count ( 99% of total)
     501 ..    1000 ===>     151 count ( 99% of total)
-   1001 ..    5000 ===>      71 count (100% of total)
+   1001 ..    5000 ===>      72 count (100% of total)
 
 ---------------------------------------------------
 Distribution of total memory used      per method (in KB):
      <=         20 ===>       0 count (  0% of total)
      21 ..      50 ===>    6051 count ( 23% of total)
      51 ..      75 ===>    8470 count ( 55% of total)
-     76 ..     100 ===>    4944 count ( 74% of total)
+     76 ..     100 ===>    4937 count ( 74% of total)
-    101 ..     150 ===>    3999 count ( 89% of total)
+    101 ..     150 ===>    3997 count ( 89% of total)
-    151 ..     250 ===>    1964 count ( 96% of total)
+    151 ..     250 ===>    1971 count ( 96% of total)
-    251 ..     500 ===>     683 count ( 99% of total)
+    251 ..     500 ===>     675 count ( 99% of total)
-    501 ..    1000 ===>     109 count ( 99% of total)
+    501 ..    1000 ===>     119 count ( 99% of total)
    1001 ..    5000 ===>      71 count (100% of total)
 Microsoft (R) CoreCLR Native Image Generator - Version 4.5.22220.0
 Copyright (c) Microsoft Corporation.  All rights reserved.
 
 Native image D:\Source\coreclr\bin\Product\Windows_NT.x64.Release\System.Private.CoreLib.dll generated successfully.
```

</details>

JosephTremoulet commented:
I've verified there are no SuperPMI desktop asm diffs for the refactoring (https://github.com/dotnet/coreclr/pull/13314/commits/99edc2b1231f94ca6d77249653d9dea47751c6ab).

BruceForstall commented:
Awesome. Thanks.

JosephTremoulet commented:
@redknightlois, this has made it into preview feeds as of [Microsoft.NETCore.App 2.1.0-preview2-25628-01](https://dotnet.myget.org/feed/dotnet-core/package/nuget/Microsoft.NETCore.App/2.1.0-preview2-25628-01) -- with `<TargetFramework>netcoreapp2.1</TargetFramework>` and `<RuntimeFrameworkVersion>2.1.0-preview2-25628-01</RuntimeFrameworkVersion>` in your .csproj and `<add key="dotnet core" value="https://dotnet.myget.org/F/dotnet-core/api/v3/index.json" />` in your NuGet.config, you can try it out.
